feat/faction-chats

abdulqaharu opened the pull request:
<!-- enter the gh issue after hash -->

- [x] issue #106 
- [x] follows contribution [guide](https://github.com/keep-starknet-strange/art-peace/blob/main/CONTRIBUTING.md)

<!-- PR description below -->

b-j-roberts commented on backend/go.mod:
```diff
@@ -20,6 +20,7 @@ require (
 	github.com/dgryski/go-rendezvous v0.0.0-20200823014737-9f7001d12a5f // indirect
 	github.com/fxamacker/cbor/v2 v2.5.0 // indirect
 	github.com/georgysavva/scany/v2 v2.1.3 // indirect
+	github.com/golang-jwt/jwt/v5 v5.2.1 // indirect
```
Is this being used anywhere? If not we shouldn't add it for now.

b-j-roberts commented on backend/routes/quests.go:
```diff
@@ -22,6 +23,20 @@ type MainQuest struct {
+type FactionChat struct {
+	ID         int       `json:"id"`
+	Sender     string    `json:"sender"`
+	FactionKey int       `json:"faction_key"`
```
Please use camel case for json fields.

b-j-roberts commented on backend/routes/quests.go:
```diff
@@ -103,3 +118,100 @@ func GetTodayStartTime(w http.ResponseWriter, r *http.Request) {
+	}
+
+	var chat FactionChat
+	err := json.NewDecoder(r.Body).Decode(&chat)
```
The body should just contain the factionKey and message, so other fields can be removed from the `FactionChat` type here.

Also, please use the `ReadJsonBody` utility from `routes/requests.go`

b-j-roberts commented on backend/routes/quests.go:
```diff
@@ -103,3 +118,100 @@ func GetTodayStartTime(w http.ResponseWriter, r *http.Request) {
 }
+
+func SendChat(w http.ResponseWriter, r *http.Request) {
+	userAddress := r.URL.Query().Get("address")
```
There will be no query params like this for the post requests.

b-j-roberts commented on backend/routes/quests.go:
```diff
@@ -103,3 +118,100 @@ func GetTodayStartTime(w http.ResponseWriter, r *http.Request) {
+	}
+	defer r.Body.Close()
+
+	err = core.PostgresExec(r.Context(), "INSERT INTO FactionChats (sender, faction_key, message) VALUES ($1, $2, $3)", chat.Sender, chat.FactionKey, chat.Message)
```
The sender can be set to a hardcoded address for now ( 0 is fine ), also only insert into factionchats if the sender is in the corresponding faction.

Please do all postgres queries / execs in a single request as well to prevent back and forth with the postgres server. There are various ways to do conditional inserts within a single query with sql

b-j-roberts commented on backend/routes/quests.go:
```diff
@@ -103,3 +118,100 @@ func GetTodayStartTime(w http.ResponseWriter, r *http.Request) {
+}
+
+func DeleteChat(w http.ResponseWriter, r *http.Request) {
+	userAddress := r.URL.Query().Get("address")
```
There will be no query params like this for the post requests.

b-j-roberts commented on backend/routes/quests.go:
```diff
@@ -103,3 +118,100 @@ func GetTodayStartTime(w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	chatID := r.URL.Query().Get("id")
```
Instead use a request body to send this id value. And use `ReadJsonBody`

b-j-roberts commented on backend/routes/quests.go:
```diff
@@ -103,3 +118,100 @@ func GetTodayStartTime(w http.ResponseWriter, r *http.Request) {
+
+	canDelete := false
+
+	leader, err := core.PostgresQueryOne[string]("SELECT ftns.leader FROM Factions ftns INNER JOIN FactionChats ftcs ON ftns.faction_key = ftcs.key WHERE ftcs.id = $1", chatID)
```
Please compress all of these back and forth checks into a single query.

You can use a hardcoded 0 address for the user here too for now.

b-j-roberts commented on backend/routes/quests.go:
```diff
@@ -103,3 +118,100 @@ func GetTodayStartTime(w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	chats, err := core.PostgresQueryJson[FactionChat]("SELECT id, sender, faction_key, message, time FROM FactionChats WHERE faction_key = $1 ORDER BY time ASC", factionKey)
```
Require the user be in the faction to get the chats.

b-j-roberts reviewed (changes requested):
Nice work so far. Left some comments for review

b-j-roberts commented:
Hey, will you be able to complete these changes soon?

abdulqaharu commented:
> Hey, will you be able to complete these changes soon?

Yup, I'll raise a PR over the weekend @b-j-roberts

b-j-roberts commented:
> > Hey, will you be able to complete these changes soon?
> 
> Yup, I'll raise a PR over the weekend @b-j-roberts

Please work off of this same PR, there is no need to open a new one. Otherwise, it makes reviewing harder and you lose all the comments.

For instance, the issue mentioned in this comment was never resolved, which would be easy to catch if you worked off of the same PR : 
https://github.com/keep-starknet-strange/art-peace/pull/127#discussion_r1591684351

abdulqaharu commented:
When I updated the branch I was working off with latest code from main before I started working, it had so many conflicts and was a bit verbose to resolve. So i removed all my initial commit, updated the branch with the latest in main, before adding back my changes and resolved the comments you added and pushed. I am guessing doing that forced a push to the branch and closed this PR. 
So I had to created another PR.

@b-j-roberts
